pinot-transport/src/main/java/com/linkedin/pinot/routing/HelixExternalViewBasedRouting.java

```diff
@@ -229,13 +370,144 @@ public void markDataResourceOnline(String tableName, ExternalView externalView,
     _brokerRoutingTable.remove(tableName);
-    _routingTableLastKnownZkVersionMap.remove(tableName);
+    _lastKnownExternalViewVersionMap.remove(tableName);
+    _lastKnownInstanceConfigsForTable.remove(tableName);
     _timeBoundaryService.remove(tableName);
+
+    // Remove table from all instances
+    synchronized (_tablesForInstance) {
+      for (String instanceName : _brokerRoutingTable.keySet()) {
+        Set<String> tablesForCurrentInstance = _tablesForInstance.get(instanceName);
+        if (tablesForCurrentInstance.contains(tableName)) {
+          tablesForCurrentInstance.remove(tableName);
+        }
+
+        if (tablesForCurrentInstance.isEmpty()) {
+          _lastKnownInstanceConfigs.remove(instanceName);
+        }
+      }
+    }
+  }
+
+  public void processExternalViewChange() {
+    // Get stats for all tables that we're serving
+    HelixDataAccessor helixDataAccessor = _helixManager.getHelixDataAccessor();
+    PropertyKey.Builder propertyKeyBuilder = helixDataAccessor.keyBuilder();
+    List<String> tablesServed = new ArrayList<>(_lastKnownExternalViewVersionMap.keySet());
+    List<String> externalViewPaths = new ArrayList<>(tablesServed.size());
+
+    for (String tableName : tablesServed) {
+      PropertyKey propertyKey = propertyKeyBuilder.externalView(tableName);
+      externalViewPaths.add(propertyKey.getPath());
+    }
+
+    Stat[] externalViewStats =
+        helixDataAccessor.getBaseDataAccessor().getStats(externalViewPaths, AccessOption.PERSISTENT);
+
+    // Make a list of external views that changed
+    List<String> tablesThatChanged = new ArrayList<>();
+
+    for (int i = 0; i < externalViewStats.length; i++) {
+      Stat externalViewStat = externalViewStats[i];
+      if (externalViewStat != null) {
+        String currentTableName = tablesServed.get(i);
+        int currentExternalViewVersion = externalViewStat.getVersion();
+        int lastKnownExternalViewVersion = _lastKnownExternalViewVersionMap.get(currentTableName);
+
+        if (lastKnownExternalViewVersion != currentExternalViewVersion) {
+          tablesThatChanged.add(currentTableName);
+        }
+      }
+    }
+
+    // Fetch the instance configs and update the routing tables for the tables that changed
+    if (!tablesThatChanged.isEmpty()) {
+      List<InstanceConfig> instanceConfigs = helixDataAccessor.getChildValues(propertyKeyBuilder.instanceConfigs());
+
+      for (String tableThatChanged : tablesThatChanged) {
+        // We ignore the external views given by Helix on external view change and fetch the latest version as our
+        // version of Helix (0.6.5) does not batch external view change messages.
+        ExternalView externalView = helixDataAccessor.getProperty(propertyKeyBuilder.externalView(tableThatChanged));
+        if (isRoutingTableRebuildRequired(tableThatChanged, externalView, instanceConfigs)) {
+          buildRoutingTable(tableThatChanged, externalView, instanceConfigs);
+        }
+      }
+    }
```

Comment:
And a log message here with how many tables we ended up updating (N out of M). Maybe good to have metrics also?